nulib/meadow
# Cache and index variants for controlled terms if present

# Summary 

We want to index variants for LOC controlled terms to improve search.

For example, we want to be able to enter "European War" or "WWII" in the search box and have works with controlled terms fields containing the LOC names authority `http://id.loc.gov/authorities/subjects/sh85148273` ("World War, 1939-1945") be returned. 

# Specific Changes in this PR

- Ecto migration to add `variants` field to `controlled_term_cache` table
- Code changes to cache variants in ETS and db

# Version bump required by the PR

See [Semantic Versioning 2.0.0](https://semver.org/) for help discerning which is required.

- [x] Minor (backwards compatible Elasticsearch API change)

# Steps to Test

In the dev environment, do a `mix deps.get` to fetch the new `Authoritex` release. And run `mix ecto.migrate` to update the db.

- Go to a work record and click edit on the "About this Item" tab
- Add an LOC controlled term with variants. For example add "World War, 1939-1945" for LOC Names Authority
- Wait for indexing to happen (or kick off manually)
- Pull up the work in Kibana and observe the controlled term indexed with variants. 
- Go to search page and search for "European War" or "WWII". 
- Observe that the work record you edited is returned in results. 

To update any existing works in your dev environment, in `iex` run:
- `Meadow.Data.ControlledTerms.clear!`
- `Meadow.Data.Indexer.reindex_all!`

# :rocket: Deployment Notes

**Note** - __if you check any of these boxes go to the [(always open) `main` <- `staging` PR](https://github.com/nulib/meadow/pulls) and add detailed notes and instructions to help out others who may end up deploying your changes to production__

- Backward compatible API changes
- Backwards-incompatible API changes
- [x] Requires reindex (see below)
- [x] Other specific instructions/tasks (see below)
  - Run `Meadow.Data.ControlledTerms.clear!` to clear the ETS and DB cache (ETS cache clear will happen on deploy anyways). 
  - Then run `Meadow.Data.Indexer.reindex_all!` to recache/reindex all works 

# Tested/Verified
